### ISSUE TITLE
Gem::Specification.add_spec is deprecated, what is the suggested replacement

### ISSUE THREAD

**Issue by macumber** [rubygems/rubygems#2469]
My current problem is that Gem::Specification.add_spec is deprecated but no replacement is suggested or obviously available.  My executable has a statically linked Ruby interpreter embedded in it.  We store the Ruby standard libraries and some gems as hard coded strings compiled into our executable.  We extract these strings and add them to the Gem registry.  Gem::Specification.add_spec used to work for our process, what is the suggested replacement?

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [ ] The command line `gem`
  - [X] Other - Direct use of the Gem::Specification API

Here are my current environment details:

We have a custom build of Ruby 2.5.1, currently using Bundler 14.4

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by colby-swandale** [comment]
Here is the relevant Changelog entry for the deprecation https://github.com/rubygems/rubygems/blob/master/History.txt#L689-L691

**Comment by macumber** [comment]
Thanks @colby-swandale I had seen that message.  Simply providing some notes on how to follow these directions using the Ruby API would resolve the issue for me:

> Gem::Specification.add_specs is deprecated and will be removed from version 3.0 with no replacement. To add specs, install the gem, then reset the cache.

Does this mean that I should install gems to a location in GEM_PATH and then call Gem::Specification::reset()?  After doing this Gem::Specification.each returns no specs.  How do I get Gem::Specification to find all the specs in GEM_PATH?

**Comment by bronzdoc** [comment]
@macumber have you tried `Gem::Specification._all` ?

**Comment by hsbt** [comment]
> How do I get Gem::Specification to find all the specs in GEM_PATH?

You can get them with `Gem::Specification.all`

**Comment by tijcolem** [comment]
I just hit this as `Gem::Specification.add_spec` was removed in 2.7.2. The deprecation warning indicated 3.0, but I guess it came early or I misunderstood it. 

The workaround I found was using [dirs=](https://github.com/ruby/ruby/blob/master/lib/rubygems/specification.rb#L927)  method which allows you to pass in an array of spec directories and this actions causes a reset of known gemspecs. It works well for official versioned ruby gems as these gems also get a corresponding .gemspec installed to `openstudio-gems/ruby/2.7.0/specifications/`

e.g. 

ls openstudio-gems/ruby/2.7.0/specifications/ | head
addressable-2.7.0.gemspec
ansi-1.5.0.gemspec
ast-2.4.2.gemspec
builder-3.2.4.gemspec

However, if you have non official version gems such as those defined liked this in a Gemfile

gem 'openstudio-extension', '= 0.4.0', :github => 'NREL/OpenStudio-extension-gem', :ref => '34d623f3be1bb8211d7cd273b253e4382afdfd76'

These get installed to `ruby/2.7.0/bundler/gems/` and do not come with a specifications directory like the official gems.  
This causes problems as the method [dirs=](https://github.com/ruby/ruby/blob/master/lib/rubygems/specification.rb#L927) appends "specifications" to each dir in the array.  Perhaps the method could be explicit and take in a directory as-is w/o modifying them by adding specifications.
